File: regions/system/library/typescript/audit-client/__tests__/audit-client.test.ts
import { describe, it, expect } from 'vitest';
import { BufferedAuditClient } from '../src/index.js';
import type { AuditEvent } from '../src/index.js';

function makeEvent(overrides: Partial<AuditEvent> = {}): AuditEvent {
  return {
    id: 'evt-1',
    tenantId: 'tenant-1',
    actorId: 'user-1',
    action: 'create',
    resourceType: 'document',
    resourceId: 'doc-1',
    timestamp: new Date().toISOString(),
    ...overrides,
  };
}

describe('BufferedAuditClient', () => {
  it('イベントを記録できる', async () => {
    const client = new BufferedAuditClient();
    await client.record(makeEvent());
    const events = await client.flush();
    expect(events).toHaveLength(1);
    expect(events[0].action).toBe('create');
  });

  it('flushでバッファがクリアされる', async () => {
    const client = new BufferedAuditClient();
    await client.record(makeEvent({ id: 'evt-1' }));
    await client.record(makeEvent({ id: 'evt-2' }));
    const first = await client.flush();
    expect(first).toHaveLength(2);
    const second = await client.flush();
    expect(second).toHaveLength(0);
  });

  it('複数イベントを順序通りに返す', async () => {
    const client = new BufferedAuditClient();
    await client.record(makeEvent({ action: 'create' }));
    await client.record(makeEvent({ action: 'update' }));
    await client.record(makeEvent({ action: 'delete' }));
    const events = await client.flush();
    expect(events.map((e) => e.action)).toEqual(['create', 'update', 'delete']);
  });

  it('空バッファでflushすると空配列を返す', async () => {
    const client = new BufferedAuditClient();
    const events = await client.flush();
    expect(events).toEqual([]);
  });
});

File: regions/system/library/dart/audit_client/test/audit_client_test.dart
import 'package:test/test.dart';

import 'package:k1s0_audit_client/audit_client.dart';

void main() {
  late BufferedAuditClient client;

  setUp(() {
    client = BufferedAuditClient();
  });

  AuditEvent makeEvent(String action) => AuditEvent(
        id: 'evt-1',
        tenantId: 'tenant-1',
        actorId: 'user-1',
        action: action,
        resourceType: 'document',
        resourceId: 'doc-1',
        timestamp: DateTime.now(),
      );

  group('record', () {
    test('イベントをバッファに追加すること', () async {
      await client.record(makeEvent('create'));
      final events = await client.flush();
      expect(events, hasLength(1));
      expect(events.first.action, equals('create'));
    });
  });

  group('flush', () {
    test('バッファ内の全イベントを返すこと', () async {
      await client.record(makeEvent('create'));
      await client.record(makeEvent('update'));
      final events = await client.flush();
      expect(events, hasLength(2));
    });

    test('flush 後にバッファがクリアされること', () async {
      await client.record(makeEvent('create'));
      await client.flush();
      final events = await client.flush();
      expect(events, isEmpty);
    });
  });

  group('AuditEvent', () {
    test('全フィールドを保持すること', () {
      final event = makeEvent('delete');
      expect(event.id, equals('evt-1'));
      expect(event.tenantId, equals('tenant-1'));
      expect(event.action, equals('delete'));
      expect(event.resourceType, equals('document'));
    });
  });
}
